systemtest/benchtest/expvarmetrics/metrics.go

```diff
@@ -0,0 +1,253 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package expvarmetrics
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+)
+
+type Metric int
+
+const (
+	Goroutines Metric = iota
+	Bytes
+	MemAllocs
+	MemBytes
+	RSSMemoryBytes
+	HeapAlloc
+	HeapObjects
+	NumGC
+	PauseTotalNs
+	AvailableBulkRequests
+	TotalEvents
+	ActiveEvents
+	TransactionsProcessed
+	SpansProcessed
+	MetricsProcessed
+	ErrorsProcessed
+	ErrorElasticResponses
+	ErrorOTLPTracesResponses
+	ErrorOTLPMetricsResponses
+)
+
+type AggregateStats struct {
+	First       int64
+	Last        int64
+	Min         int64
+	Max         int64
+	Mean        float64
+	sampleCount int64
+}
+
+type watchItem struct {
+	notifyChan chan<- bool
+	validator  func(int64) bool
+}
+
+// Collector defines a metric collector which queries expvar
+// endpoints and aggregates the collected metrics.
+//
+// Watchers are one-time hooks into collector to know about when
+// the a specific state of the metric is observed by the collector.
+type Collector struct {
+	l         sync.Mutex
+	serverURL string
+	period    time.Duration
+	metrics   map[Metric]*AggregateStats
+	watchers  map[Metric]*watchItem
+}
```

Comment:
I think we can simplify this structure a little bit and only store what's necessary. `serverURL` and `period` are only used once when the collector is initialized, we can just pass these to the functions that require them.

Additionally, I see no reason to pollute the heap with AggregateStats and watchItems, since it's relatively easy to access or pass around the values to be updated within the collector.

Last, `watches` seems like a better name than `watchers`.

The `closed` boolean is used later on when the watches are closed and the loop function isn't running anymore (See comments below, towards the end of the review).

```suggestion
// Watches are one-time hooks into collector to know about when
// the a specific state of the metric is observed by the collector.
type Collector struct {
	l       sync.Mutex
	metrics map[Metric]AggregateStats
	watches map[Metric]watch
	closed  bool
}
```